runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -1108,7 +1110,42 @@ function M.display(diagnostics, bufnr, client_id, config)
   if signs_opts then
     M.set_signs(diagnostics, bufnr, client_id, nil, signs_opts)
   end
+
+  -- cache extmarks
+  M.save_extmarks(bufnr, client_id)
+end
+
+-- caches the extmarks set by M.display
+function M.save_extmarks(bufnr, client_id)
+  local ns = vim.lsp.diagnostic._get_diagnostic_namespace(client_id)
+  diagnostic_cache_extmarks[bufnr][client_id] = vim.api.nvim_buf_get_extmarks(bufnr, ns, 0, -1, {details = true})
 end
+
+-- restores the extmarks set by M.display
+function M.restore_extmarks(bufnr)
+  local lcount = vim.api.nvim_buf_line_count(bufnr)
+  for client_id, extmarks in pairs(diagnostic_cache_extmarks[bufnr]) do
+    local ns = vim.lsp.diagnostic._get_diagnostic_namespace(client_id)
+    local extmarks_current = vim.api.nvim_buf_get_extmarks(bufnr, ns, 0, -1, {details = true})
+    local found = {}
+    for _, extmark in ipairs(extmarks_current) do
+      -- HACK: the missing extmakrs seems to still exist, but at the line after the last
+      if extmark[2] < lcount then
+        found[extmark[1]] = true
+      end
+    end
+    for _, extmark in ipairs(extmarks) do
+      if not found[extmark[1]] then
+        local opts = extmark[4]
+        opts.id = extmark[1]
+        -- HACK: end_row is somtimes present, but not a valid option
+        opts.end_row = nil
```

Comment:
seems we accidentally `end_line` in set_extmarks instead :P, while we can't remove it, we could allow either:

```
// Integer nvim_buf_set_extmark(Buffer buffer, Integer ns_id, ...)
...
    } else if (strequal("end_line", k.data) || strequal("end_row", k.data)) {
```

Reply:
those are options coming from get_extmarks, so maybe fix it there?

Reply:
Just pushed a fix that sets it to `end_line` instead, if `end_row` is present

Reply:
that would break the interface of `get_extmarks`. but this is fine for now, we can fix the general extmark issue in a separate PR.